internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+type LogManager interface {
+	PlanChanged(*plan.Plan)
+	NewService(serviceName string, buffer *servicelog.RingBuffer)
+	StopService(serviceName string)
+}
+
+type logManager struct {
+	// Store services and selected log targets, as defined by plan
+	planTargets serviceTargetSet
+	// Store buffers for each running service
+	buffers map[string]*servicelog.RingBuffer
+	// Store log forwarders for each service & target
+	forwarders map[serviceTarget]logForwarder
+	// Log forwarders to be added once the service starts
+	toAdd serviceTargetSet
+}
+
+func NewLogManager() LogManager {
+	return &logManager{
+		buffers:    map[string]*servicelog.RingBuffer{},
+		forwarders: map[serviceTarget]logForwarder{},
+		toAdd:      serviceTargetSet{},
+	}
+}
+
+func (m *logManager) PlanChanged(p *plan.Plan) {
+	m.selectTargets(p)
```

Comment:
I would prefer the naive approach here, if the plan changes, stop all the forwarders (let them flush), start all the forwarders again (using the same iterator so they can resume from where they left off)

Reply:
not sure I like this, I'd rather keep them running if nothing changes. @benhoyt what do you think?